DeeperFunc/Path.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  Path.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 27/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+func lit(_ str: String) -> RoutePattern<Void, Path> {
+    return .init(
+        parse: { route in
+            guard route.path.first == str else { return nil }
+            return (RouteComponents(path: Array(route.path.dropFirst()), query: route.1), ())
+    },
+        print: { _ in
+            return RouteComponents(path: [str], query: [:])
+    }, template: str)
+}
+
+func pathParam<A>(_ iso: PartialIso<String, A>) -> RoutePattern<A, Path> {
+    return .init(
+        parse: { route in
+            guard let pathComponent = route.path.first?.removingPercentEncoding, let parsed = iso.apply(pathComponent) else { return nil }
+            return ((Array(route.path.dropFirst()), route.1), parsed)
+    },
+        print: { a in
+            return RouteComponents(path: [iso.unapply(a)].flatMap({ $0 }), query: [:])
+    }, template: pathParamTemplate(A.self))
+}
+
+// These are params transformations
+
+public let string: RoutePattern<String, Path> = pathParam(.id)
+public let int: RoutePattern<Int, Path> = pathParam(.int)
+public let double: RoutePattern<Double, Path> = pathParam(.double)
+
+// drop left param
+infix operator /> : MultiplicationPrecedence
```

Comment:
not fond of all these custom operators, but using `/` for all cases makes expressions to complex for swift compiler